{% extends "base.html" %}

{% block title %} PRECOGx | About  {% endblock %}

{% block content %}
<section class="pt-xxl-8" id="home">
    <div class="container">
        <div class="row align-items-center">
            <div class="col-md-3 order-lg-1 mb-5 mb-lg-0"></div>
            <div class="col-md-12 text-center text-md-start fs-1 mb-5">
                <h1 class="fw-bold mb-4">About - Methods</h1>
                <h2 class="fw-bold mb-4">Embeddings generations</h2>
                <p class="text-justify fs-0">Embeddings of the protein sequences were generated by using pre-trained protein language models that have been recently released [REF].
                    We computed embeddings from fasta sequence using the extract.py function of the esm library (https://github.com/facebookresearch/esm)
                    and by specifying the ESM-1b model (esm1b_t33_650M_UR50S) with embedding for individual amino acids as well as averaged over the full
                    sequence using the option <i>“--include mean per_tok”</i>.
                </p>
                <p class="text-justify fs-0">We generated embeddings for each individual layers separately, including the final one, by specifying their corresponding
                    number in the <i>“--repr-layers” option</i>.
                </p>
                <h2>Data sets</h2>
                <p class="text-justify fs-0">We used experimental binding affinities from two distinct sources:  TGFα shedding assay[REF],
                    entailing affinities of 148 GPCRs for 11 chimeric G proteins, and the eBRET assay[REF], providing affinities of 97
                    receptors for 12 G-proteins and 3 β-arrestins/GRKs binders. For the TGFα shedding assay, we considered a receptor to
                    be coupled to a certain G-proteins if the logarithm of the relative intrinsic activity (logRAi) was greater than -1. Likewise,
                    in the eBRET assay we considered a receptor to be coupled to a G-protein (or β-arrestins/GRK) if the binding efficacy
                    (Emax) was greater than  0.
                </p>
                <p class="text-justify fs-0">We also considered a unified coupling dataset derived from a meta-analysis,
                    consisting on standardized normalization and consensus support, of these two large-scale,
                    quantitative binding assays [REF].
                </p>
                <img class="pt-md-0 center" src="static/img/gallery/workflow.png" alt="Method workflow"/>
                <h2 class="fw-bold mb-4">Model training</h2>
                <p class="text-justify fs-0">We developed the new PRECOGX by training multiple models by using as independent set of features each of the layers,
                    including the final one, from the protein embeddings obtained from the pre-trained ESM-1b. For each G-protein, we defined a training set
                    by labelling receptors with reported binding greater than the cutoff (see above) as coupled, and the remaining ones as not coupled.
                    We implemented the predictor using either a logistic regression or support vector classifier (https://scikit-learn.org/).
                    A grid search was performed over a stratified 5-fold cross validation (CV) to select the best parameters of the algorithm,
                    which we repeated 10 times to ensure minimum variance. We generated a total of 34 models per Gprotein ( or β-arrestin) and assay.
                    The best models where chosen based on highest AUC score in the 5-fold CV while the |Recall(REC) - Specificty(SPE)| was below 0.15.
                    Finally we considered the 17 best performing models for each G-protein or β-arrestin.
                </p>
                <h2 class="fw-bold mb-4">Model testing</h2>
                <p class="text-justify fs-0">We benchmarked our method against PRECOG, the web-server for GPCR- G-protein coupling predictions that we
                    previously developed [REF]. We obtained a list of 80 (TGFα shedding assay as training set) or 135 (ebBret assay as training set)
                    unseen GPCRs from GtPDB (http://www.guidetopharmacology.org/) that are present in neither ours nor PRECOG training set. Since IUPHAR
                    lacks a proper true negative set, we used REC to compare the performance of our predictor with Pred-Couple’s one. In order to assess
                    over-fitting, we performed the randomization test by replacing the original labels of the training matrix with randomly assigned labels,
                    while preserving the ratio of number of positive (coupling) and negative (uncoupling) GPCRs.
                </p>
                <h2 class="fw-bold mb-4">PCA of the GPCRome embedded space</h2>
                <p class="text-justify fs-0">We generated embeddings for the human GPCRome, comprising a total of 377 receptors (279 Class A, 33 Class B,
                    17 class C, 11 class F, 25 Taste receptors and 14 in other classes). We considered either the embeddings generated by considering the final,
                    33th layer, or the layer giving best performances during training. Embeddings were subjected to Principal Component Analysis, using the PCA
                    function from sklearn.decomposition. Each GPCR sequence within the embedding was annotated via coupling information from either TGFα shedding,
                    eBRET, or GtPdb or STRING (for B-arrs).
                </p>
                <h2 class="fw-bold mb-4">Contact analysis</h2>
                <p class="text-justify fs-0">In order to interpret the determinants of binding specificity, we first calculated predicted contacts for each
                    sequence using a logistic regression over the model's attention maps [REF], available in the esm library through the predict_contacts function.
                    Then, the predicted contact maps were grouped on the basis of G-protein binding specificity and contrasted to the contact maps of all the
                    GPCRs which was used as a background. We computed a differential contact maps by calculating a log-odds ratio, employing the following formula:
                </p>
                <img class="pt-md-0 center" src="static/img/gallery/formula.png" alt="Math formula"/>
                <p class="text-justify fs-0">Where first term represents a ratio where in numerator is the ratio of a number of GPCRs coupled to specific
                    Gprotein for a specific contact pair and the total number of GPCRs coupled to specific Gprotein depending on the assay considered.
                    In the denominator is a ratio of total number of GPCRs having a contact pair and the total number of GPCR in the specific assay.
                    The second term represents a ratio where in numerator is the ratio of a specific contact pair count across GPCRs coupled to specific
                    Gprotein and the total number of GPCRs coupled to specific Gprotein depending on the assay considered. Denominator is a ratio of a specific
                    contact pair count across all GPCRs and the total number of GPCR in the specific assay. Contact pairs considered are those with a
                    probability higher than 0.5 calculated based on predict_contacts function and those appearing in at least 15% of all GPCR in the assays.
                    Contacts were normalized using the maxabs scaler from scikit-learn.
                </p>
                <p class="text-justify fs-0">Contacts with a positive log-odds ratio (enriched) are seen more frequently in receptors coupled to a specific G-protein, while contacts with
                    a negative log-odds ratio (depleted) are seen less frequently in receptors coupled to a specific G-protein.
                </p>
                <img class="pt-md-0 center" src="static/img/gallery/contact.png" alt="contact analysis"/>

                <h5 class="fw-bold mb-4">Contacts</h5>
                <p class="text-justify fs-0">Francesco Raimondi - francesco.raimondi@sns.it</p>
                <p class="text-justify fs-0">Marin Matic - marin.matic@sns.it</p>
                <p class="text-justify fs-0">Gurdeep Sing - gurdeep.sing@bioquant.uni-heidelberg.de</p>
            </div>
        </div>
    </div>
</section>

{% endblock%}