{% extends "base.html" %}

{% block title %} PRECOGx | Help  {% endblock %}

{% block content %}
<section class="pt-xxl-8" id="home">
    <div class="container">
        <div class="row align-items-center">
            <div class="col-md-3 order-lg-1 mb-5 mb-lg-0"></div>
            <div class="col-md-12 text-center text-md-start fs-1 mb-5">
                <h1 class="fw-bold mb-4">Help</h1>
                <h2 class="fw-bold mb-4"> Web-server usage</h2>
                <p class="text-justify fs-0">
			A user can input either a wild type (WT) or mutated protein sequence. In the latter case, calculation of the WT is also carried out for comparison purposes. Protein embeddings from the pre-trained <a href="https://github.com/facebookresearch/esm" target="_blank">ESM-1b model</a> are calculated for the input sequence. These are used to (i) calculate the coupling preferences of the input to G-protein and β-arrestins  (ii) project them on an embedded space encompassing the human GPCRome, and (iii) visualise predicted contacts enriched/depleted for a given coupling group in 1D (sequence), 2D (contact map) or 3D (structure).
                </p>
                <h2 class="fw-bold mb-4">Input format and options</h2>
                <p class="text-justify fs-0">
		PRECOGx requires a list of protein identifiers (<a href="https://www.uniprot.org/" target="_blank">UniProt</a> Accessions, Entry names or Gene symbols), whole sequence in FASTA format, or mutations, by simply appending the positional information (e.g. D294H), separated by a forward slash (/), to the protein. On the forms there are some additional options:

                </p>
                <ul>
			<li class="text-justify fs-0"><strong>UniProt ACC/Entry names/Gene symbols</strong>: Set of proteins specified using the UniProt <a href="https://www.uniprot.org/help/accession_numbers" target="_blank">accessions</a>, <a href="https://www.uniprot.org/help/entry_name" target="_blank">Entry names</a> and <a href="https://www.uniprot.org/help/gene_name" target="_blank">Gene symbols</a></li>
		    <li class="text-justify fs-0"><strong>Mutations</strong>: Set of mutations</li>
		    <li class="text-justify fs-0"><strong>FASTA format</strong>: A protein specified as <a href = "https://en.wikipedia.org/wiki/FASTA_format" target="_blank">whole sequence</a></li>
                    <li class="text-justify fs-0"><strong>Submit</strong>: Execute the input</li>
                </ul>
		<p class="text-justify fs-0">
		See <a href="https://precogx.bioinfolab.sns.it" target="_blank">examples</a> on home page.
		</p>
                <h2 class="fw-bold mb-4">Output Page</h2>
                <img class="pt-md-0 center" src="static/img/gallery/output_page.png" alt="result page"/>
                <p></p>

                <h2 class="fw-bold mb-4">Coupling probabilities panel</h2>
                <div class="row min-vh-xl-100 min-vh-xxl-25">
                    <div class="col-md-6">
                        <img class="pt-md-0 center w-100" src="static/img/gallery/coupling_probabilities.png" alt="result page"/>
                    </div>
                    <div class="col-6">
                        <p class="text-justify fs-0">
			The table here represents coupling probabilities predicted by PRECOGx as well experimental information provided by <a href ="https://www.guidetopharmacology.org/" target="_blank">GtoPdb/IUPHAR</a>, the <a href="https://pubmed.ncbi.nlm.nih.gov/31160049/" target="_blank">TGFɑ shedding</a> and the <a href="https://www.biorxiv.org/content/10.1101/2020.04.20.052027v1" target="_blank">ebBRET</a> assays.
			</p>

                        <p class="text-justify fs-0">Columns:</p>
                        <ul>
                            <li class="text-justify fs-0"><strong>GPCR</strong>: Input name(s) provided by the user.</li>
                            <li class="text-justify fs-0"><strong>VAR</strong>: Type of variants - (a) Wild type ;(b) Mutation ;(c) IUPHAR;(d) LogRAi;(e) Emax
</li>
                            <li class="text-justify fs-0"><strong>[GNAS - β-arr/GRK2]</strong>: Coupling probability predicted by PRECOGx (for wild type and mutations) for the G-proteins and β-arrestins instances available;</li>
                        </ul>
                        <p class="text-justify fs-0">Rows with:</p>
                        <ul>
				<li class="text-justify fs-0"><strong>VAR = IUPHAR</strong> display information known about the given GPCR known from GtPdb/IUPHAR.</li>
        	                <li class="text-justify fs-0"><strong>VAR = LogRAi</strong> display information known about the input GPCR from the <strong>TGFɑ shedding assay</strong> experiment.</li>
                	        <li class="text-justify fs-0"><strong>VAR = Emax</strong> display information known about the input GPCR from the <strong>ebBRET assay</strong>.</li>
                        	<li class="text-justify fs-0"><strong>VAR = WT</strong> display coupling probabilities of the given GPCR predicted by PRECOGx.</li>
	                        <li class="text-justify fs-0"><strong>VAR = Mutation</strong> (eg: D294H - Aspartic acid at sequence position 294 mutated to Histidine) displays coupling probabilities of the input GPCR's mutation predicted by PRECOGx.</li>
                        </ul>
			<p class="text-justify fs-0">
			<strong>Click</strong> cells with  VAR = WT or Mutation to view the input GPCR in the PCA, Sequence/Predicted contacts and Structure panels (see below). By default information related to the coupling with  highest probability will be displayed in the panels. To visualise information related to other G-protein/β-arrestin couplings, simply click the corresponding cell in the table.
			</p>
                    </div>
                </div>


                <h2 class="fw-bold mb-4">PCA panel</h2>
                <div class="row min-vh-xl-100 min-vh-xxl-25">
                    <div class="col-md-6">
                        <img class="pt-md-0 center w-100" src="static/img/gallery/PCA.png" alt="result page"/>
                    </div>
                    <div class="col-6">
                        <p class="text-justify fs-0">
			The PCA panel displays the projections of the human GPCRome sequences on the first two eigenvectors (principal components) obtained by performing the Principal Component Analysis (PCA) on the embeddings of the human GPCRome or the best layer on which the given G-protein was trained. The user is also provided the option to annotate the human GPCRome based on their known functional information with the help of a drop-down menu. If either the shedding or ebBRET assay is specified in the drop down, a continuous colour gradient, ranging from <span style="background-color: silver;">dark grey</span> (no binding) to <span style="background-color: lightgreen;">green</span> (maximum binding), will be applied. If GtoPdb/IUPHAR is selected, a discrete color  will be applied: <span style="background-color: darkseagreen;">dark green</span> (primary transduction mechanisms) and <span style="background-color: lightgreen;">pale green</span> (secondary transduction mechanisms). Wild type and mutated (if provided) sequences are highlighted in, <span style="background-color: khaki;">khaki</span> and <span style="background-color: violet;">violet</span>, respectively.

                        </p>
                    </div>
                </div>
                <p></p>

                <h2 class="fw-bold mb-4">Predicted contact Panel</h2>
                <div class="row min-vh-xl-100 min-vh-xxl-25">
                    <div class="col-md-6">
                        <img class="pt-md-0 center w-100" src="static/img/gallery/predicted_contacts.png" alt="result page"/>
                    </div>
                    <div class="col-6">
                        <p class="text-justify fs-0">
			The Predicted contact panel displays differential predicted contacts for a particular coupling group. The predicted contacts for each individual receptor are grouped together on the basis of a specific coupling preferences and summarised via  log-odds ratio statistics, indicating contacts that are more (shown in green) or less (red) frequently than expected. Contact statistics is referenced to <a href = "https://www.sciencedirect.com/science/article/abs/pii/S1043947105800497" target="_blank">Ballesteros-Weinstein (BW) numbering scheme</a> obtained from <a href="https://gpcrdb.org/" target="blank_">GPCRDB</a>. The user will also find a slider widget, which can be used filter and visualize the predicted contacts with log-odds ratio greater than user-specified absolute value.
                        </p>
                    </div>
                    <p></p>
                    <h2 class="fw-bold mb-4">Sequence Panel</h2>
                    <div class="row min-vh-xl-100 min-vh-xxl-25">
                        <div class="col-md-6">
                            <img class="pt-md-0 center w-100" src="static/img/gallery/sequence_panel.png" alt="result page"/>
                        </div>
                        <div class="col-6">
                            <p class="text-justify fs-0">
			    The sequence panel shows the positions mediating predicted contacts (highlighted in <span style="background-color: yellow;">____</span>) for the given   coupling group, which can selected by clicking on any cell in the Coupling probabilities table  (see above). The user can hover over any highlighted position to view its BW numbering. Contacts  to be projected in the sequence can be varied using the slider. Mutations, if specified, are highlighted in <span style="background-color: violet;">____</span>.
                            </p>
                        </div>
                    </div>
                    <p></p>
                    <h2 class="fw-bold mb-4">Structure Panel</h2>
                    <div class="row min-vh-xl-100 min-vh-xxl-25">
                        <div class="col-md-6">
                            <img class="pt-md-0 center w-100" src="static/img/gallery/structure_panel.png" alt="result page"/>
                        </div>
                        <div class="col-6">
                            <p class="text-justify fs-0">
			    This panel shows the differential predicted contacts of the input sequence on a 3D structure of GPCR - G-protein/β-arrestin complex. Predicted contact positions are highlighted as <span style="background-color: yellow;">spheres</span>, whose diameter is proportional to the number of mediated predicted contacts. Labels in <span style="background-color: black;">____</span> indicate corresponding BW numbering for each predicted contact. <span style="background-color: lightgreen;">Green</span> and <span style="background-color: lightcoral;">red</span> edges indicate predicted contact pairs that enriched or depleted, respectively, for a the given coupling group. If provided, the mutation is shown as a <span style="background-color: violet;">sphere</span>. The 3D structure with highest sequence identity to the input GPCR (using PSI-BLAST) is chosen and displayed by default. The user can also choose another 3D  structure from the dropdown menu, which has the list of 3D structures (PDB-ID) in descending order of sequence identity with the input GPCR.  The user will also find a slider, which can be used to filter and visualize the predicted contacts with log-odds ratio greater than user-specified absolute value. Additional toggle options are available within the NGL viewer.
                            </p>
                        </div>
                    </div>

                    <h2 class="fw-bold mb-4">Download</h2>
                    <p class="text-justify fs-0">
                        Download is available as TSV format. This gives information of the input and prediction probabilties by PRECOGx.
                    </p>
                    <p class="text-justify fs-0"> Column information: </p>
		    <ul>
                            <li class="text-justify fs-0"><strong>GPCR</strong>: Input name(s) provided by the user.</li>
                            <li class="text-justify fs-0"><strong>VAR</strong>: Type of variants - (a) Wild type ;(b) Mutation ;(c) IUPHAR ;(d) LogRAi;(e) Emax
</li>
                            <li class="text-justify fs-0"><strong>[GNAS - β-arr/GRK2]</strong>: Coupling probability predicted by PRECOGx (for wild type and mutations) for the G-proteins and β-arrestins instances available;</li>
                        </ul>

                        <p class="text-justify fs-0">Rows with:</p>
			<ul>
				<li class="text-justify fs-0"><strong>VAR = IUPHAR</strong> display information known about the given GPCR known from GtPdb/IUPHAR.</li>
                	        <li class="text-justify fs-0"><strong>VAR = LogRAi</strong> display information known about the input GPCR from the <strong>TGFɑ shedding assay</strong> experiment.</li>
        	                <li class="text-justify fs-0"><strong>VAR = Emax</strong> display information known about the input GPCR from the <strong>ebBRET assay</strong>.</li>
	                        <li class="text-justify fs-0"><strong>VAR = WT</strong> display coupling probabilities of the given GPCR predicted by PRECOGx.</li>
                        	<li class="text-justify fs-0"><strong>VAR = Mutation</strong> (eg: D294H - Aspartic acid at sequence position 294 mutated to Histidine) displays coupling probabilities of the input GPCR's mutation predicted by PRECOGx.</li>
                    </ul>
                    <h2 class="fw-bold mb-4">Browser compatibility</h2>
                    <table class="table table-bordered">
                        <thead>
                        <th>OS</th>
                        <th>Version</th>
                        <th>Chrome</th>
                        <th>Firefox</th>
                        <th>Safari</th>
                        </thead>
                        <tbody>
                        <tr class="text-justify fs-0">
                            <td>Linux</td>    <td>Ubuntu 20.04.3 LTS</td>    <td>not tested</td>    <td>95.0</td>    <td>n/a</td>
                        </tr>
                        <tr class="text-justify fs-0">
                            <td>Linux</td>    <td>CentOS 7.2.1511</td>  <td>not tested</td>    <td>ESR 52.6.0</td>    <td>n/a</td>
                        </tr>
                        <tr class="text-justify fs-0">
                            <td>Windows</td>    <td>10</td>    <td>v96.0.4664.110</td>    <td>n/a</td>   <td></td>
                        </tr>
                        <tr class="text-justify fs-0">
                            <td>MacOS</td>    <td>BigSur v11.5.2</td>    <td>v96.0.4664.110</td>    <td>v95.0.2</td>    <td>14.1.2</td>
                        </tr>
                        <tr class="text-justify fs-0">
                            <td>MacOS</td>    <td>BigSur v11.6</td>    <td>v96.0.4664.110</td>    <td>v95.0.2</td>    <td>14.1.2</td>
                        </tr>
                        </tbody>
                    </table>
                    <p></p>
                    <h5 class="fw-bold mb-4">Contacts</h5>
                    <p class="text-justify fs-0">Francesco Raimondi - francesco.raimondi@sns.it</p>
                    <p class="text-justify fs-0">Marin Matic - marin.matic@sns.it</p>
                    <p class="text-justify fs-0">Gurdeep Singh - gurdeep.singh@bioquant.uni-heidelberg.de</p>
                </div>
            </div>
        </div>
</section>
{% endblock%}
