{% extends "base.html" %}

{% block title %} PRECOGx | Help  {% endblock %}

{% block content %}
<section class="pt-xxl-8" id="home">
    <div class="container">
        <div class="row align-items-center">
            <div class="col-md-3 order-lg-1 mb-5 mb-lg-0"></div>
            <div class="col-md-12 text-center text-md-start fs-1 mb-5">
                <h1 class="fw-bold mb-4">Help</h1>
                <h2 class="fw-bold mb-4"> Web-server usage</h2>
                <p class="text-justify fs-0">A user can input either a wild type (WT) or mutated protein sequence or identifier.
                    In the latter case, calculation of the WT is also carried out for comparison purposes. Protein embeddings from
                    the pre-trained ESM-1b model are calculated for the input sequence. These are used to calculate the G-protein and
                    β-arrestin binding preferences for the input sequences, to project them on an embedded space encompassing the human GPCRome,
                    as well as to visualize predicted contacts enriched or depleted for a given coupling group in 1D (sequence), 2D (contact map)
                    or 3D by mapping them on available complex structures.
                 </p>
                 <h2 class="fw-bold mb-4">Input format and options</h2>
                 <p class="text-justify fs-0">Precogx requires a list of protein identifiers (UniProt Accession or Gene symbols),
                     whole sequence in FASTA format, or mutations, by simply appending the positional information (e.g. D154R),
                     separated by a forward slash (/), to the protein ID. On the forms there are some additional options:
                </p>
                <ul>
                    <li class="text-justify fs-0">UniProt ID/Gene symbols: Set of proteins specified using the UniProt accessions and Gene symbols</li>
                    <li class="text-justify fs-0">Mutations: Set of mutations</li>
                    <li class="text-justify fs-0">FASTA format: A protein specified as whole sequence</li>
                    <li class="text-justify fs-0">Submit: Execute the entries</li>
                </ul>
                <h2 class="fw-bold mb-4">Output Page</h2>
                <img class="pt-md-0 center" src="static/img/gallery/result.png" alt="result page"/>
                <p></p>
                 <h2 class="fw-bold mb-4">Coupling probabilities panel</h2>
                <img class="pt-md-0 center" src="static/img/gallery/coupli.png" alt="result page"/>
                <p class="text-justify fs-0">The table here represents coupling probabilities predicted by PRECOG as well information known from IUPHAR and the Shedding Assay experiment.</p>
                <p class="text-justify fs-0">Columns:</p>
                <ul>
                    <li class="text-justify fs-0">GPCR Input name(s) provided by the user.</li>
                    <li class="text-justify fs-0">VAR Type of variants - (a)Wild type P(WT); (b)Mutation P(Mutation); (c) IUPHAR; (d)LogRAI; (e) Emax</li>
                    <li class="text-justify fs-0">Coupling probabilties predicted by PRECOGX (for wild type and mutations) for the G-proteins and β-arrestin instances available;</li>
                </ul>
                <p class="text-justify fs-0">Rows:</p>
                <ul>
                    <li class="text-justify fs-0">VAR = IUPHAR display information known about the given GPCR known from GtPdb/IUPHAR.</li>
                    <li class="text-justify fs-0">VAR = LogRAI display information known about the given GPCR from the Shedding Assay experiment. The coupling values (greater than -1.0) are shown in green while the uncoupling ones are shown in red.</li>
                    <li class="text-justify fs-0">VAR = Emax display information known about the given GPCR from the ebBRET assay. </li>
                </ul>

            </div>
        </div>
    </div>
</section>
{% endblock%}
