{% extends "base.html" %}

{% block title %} PRECOGx | Help  {% endblock %}

{% block content %}
<section class="pt-xxl-8" id="home">
    <div class="container">
        <div class="row align-items-center">
            <div class="col-md-3 order-lg-1 mb-5 mb-lg-0"></div>
            <div class="col-md-12 text-center text-md-start fs-1 mb-5">
                <h1 class="fw-bold mb-4">Help</h1>
                <h2 class="fw-bold mb-4"> Web-server usage</h2>
                <p class="text-justify fs-0">A user can input either a wild type (WT) or mutated protein sequence or identifier.
                    In the latter case, calculation of the WT is also carried out for comparison purposes. Protein embeddings from
                    the pre-trained ESM-1b model are calculated for the input sequence. These are used to calculate the G-protein and
                    β-arrestin binding preferences for the input sequences, to project them on an embedded space encompassing the human GPCRome,
                    as well as to visualize predicted contacts enriched or depleted for a given coupling group in 1D (sequence), 2D (contact map)
                    or 3D by mapping them on available complex structures.
                </p>
                <h2 class="fw-bold mb-4">Input format and options</h2>
                <p class="text-justify fs-0">Precogx requires a list of protein identifiers (UniProt Accession or Gene symbols),
                    whole sequence in FASTA format, or mutations, by simply appending the positional information (e.g. D154R),
                    separated by a forward slash (/), to the protein ID. On the forms there are some additional options:
                </p>
                <ul>
                    <li class="text-justify fs-0">UniProt ID/Gene symbols: Set of proteins specified using the UniProt accessions and Gene symbols</li>
                    <li class="text-justify fs-0">Mutations: Set of mutations</li>
                    <li class="text-justify fs-0">FASTA format: A protein specified as whole sequence</li>
                    <li class="text-justify fs-0">Submit: Execute the entries</li>
                </ul>
                <h2 class="fw-bold mb-4">Output Page</h2>
                <img class="pt-md-0 center" src="static/img/gallery/result.png" alt="result page"/>
                <p></p>
                <h2 class="fw-bold mb-4">Coupling probabilities panel</h2>
                <img class="pt-md-0 center" src="static/img/gallery/coupli.png" alt="result page"/>
                <p class="text-justify fs-0">The table here represents coupling probabilities predicted by PRECOG as well information known from IUPHAR and the Shedding Assay experiment.</p>
                <p class="text-justify fs-0">Columns:</p>
                <ul>
                    <li class="text-justify fs-0">GPCR Input name(s) provided by the user.</li>
                    <li class="text-justify fs-0">VAR Type of variants - (a)Wild type P(WT); (b)Mutation P(Mutation); (c) IUPHAR; (d)LogRAI; (e) Emax</li>
                    <li class="text-justify fs-0">Coupling probabilties predicted by PRECOGX (for wild type and mutations) for the G-proteins and β-arrestin instances available;</li>
                </ul>
                <p class="text-justify fs-0">Rows:</p>
                <ul>
                    <li class="text-justify fs-0">VAR = IUPHAR display information known about the given GPCR known from GtPdb/IUPHAR.</li>
                    <li class="text-justify fs-0">VAR = LogRAI display information known about the given GPCR from the Shedding Assay experiment. The coupling values (greater than -1.0) are shown in green while the uncoupling ones are shown in red.</li>
                    <li class="text-justify fs-0">VAR = Emax display information known about the given GPCR from the ebBRET assay.</li>
                    <li class="text-justify fs-0">VAR = P(WT) display coupling probabilites of the given GPCR predicted by PRECOGX.</li>
                    <li class="text-justify fs-0">VAR = P(Mutation) (eg: Y90W - Tyrosine at sequence position 90 mutated to Tryptophan) displays coupling probabilites
                        of the given GPCR's mutation predicted by PRECOGX. The coupling probabilites
                        (greater than 0.5) are shown in green while the uncoupling ones (less than 0.5) are shown in red.</li>
                </ul>
                <p class="text-justify fs-0">Click on probabilities in VAR = P(WT/Mutation) to view the GPCR in the PCA, Sequence and Structure panels.
                    By default information relating to the coupling with the highest probability will be displayed in the
                    sequence/contact panel as well as in the 3D structure one. It is possible to visualize features related
                    to other G-protein (or β-arrestin)couplings by simply clicking on the corresponding cell of the results table.
                    This action will automatically update the features highlighted in the sequence/contact and structure panel.
                </p>
                <h2 class="fw-bold mb-4">PCA panel</h2>
                <img class="pt-md-0 center" src="static/img/gallery/pca.png" alt="result page"/>
                <p class="text-justify fs-0">The PCA panel displays the projections of the human GPCRome
                    sequences on the first two Principal Components obtained by performing the PCA on the embeddings
                    of the human GPCRome. It is possible to color annotate the human GPCRs in this panel via a drop-down
                    menu on the basis of coupling mechanisms or class/family classification. If either Shedding or eBRET is
                    specified, a continuous color gradient, ranging from gray (no binding) to green (maximum binding), will be applied.
                    If IUPHAR coupling is selected, a discrete color  will be applied: dark green (primary transduction mechanisms) and pale green
                    (secondary transduction mechanisms).  The input sequences are also displayed and annotated with a different color (e.g. fucsia).
                </p>
                <h2 class="fw-bold mb-4">Predicted contact Panel</h2>
                <img class="pt-md-0 center" src="static/img/gallery/contacts.png" alt="result page"/>
                <p class="text-justify fs-0">The Predicted Contact panel displays differential predicted contacts for a particular
                    coupling group. The predicted contacts for each individual receptor are pooled together on the basisi
                    of a specific coupling preferences and summarized via a log-odds ratio statistics, indicating contacts
                    that are more (colored in greed) or less (red) frequently than expected. Contact statistics is referenced
                    to Ballesteros-Weinstein numbering scheme obtained from GPCRDB (https://gpcrdb.org/). A filter cutoff
                    widget allows to visualize contacts with log-odds ratio greater than a certain absolute value threshold.
                </p>
                <img class="pt-md-0 center" src="static/img/gallery/sequence.png" alt="result page"/>
                <p class="text-justify fs-0">The sequence panel shows in 1D at the level of the input GPCR
                    sequence the positions (in yellow) mediating predicted contacts for a certain coupling group,
                    selected by clicking on either the “GPCR” or “VAR” fields or any of the predicted coupling probabilites
                    in the Results panel (see above). Input mutations are highlighted in magenta. Contacts projected in the
                    sequence are selected through the filter cutoff. Click a significant position to view it on a structure
                    in the Structure panel and to view its amino acid composition in a modal.
                </p>
                <h2 class="fw-bold mb-4">Structure Panel</h2>
                <img class="pt-md-0 center" src="static/img/gallery/structure.png" alt="result page"/>
                <p class="text-justify fs-0">This panel shows in 3D the differential predicted
                    contacts displayed in the Predicted contact panel. Positions are highlighted as
                    yellow spheres, whose diameter is proportional to the number of mediated predicted contacts.
                    Black labels indicate corresponding BW numbering for each sphere.
                    Green and red edges indicate predicted contact respectively enriched or depleted for a
                    specific coupling group. Whenever present in the input,  mutations are represented as a magenta sphere.
                    The PDB structure that best matches to the given GPCR (based on sequence identity using PSI-BLAST) is chosen and displayed.
                    However, the user has the option of choosing a different PDB structure from the dropdown menu (PDB-ID).
                    Several center and toggle options are available within the NGL viewer.
                </p>
                <h2 class="fw-bold mb-4">Download</h2>
                <p class="text-justify fs-0">
                    Download is available as TSV format. This gives information of the input and prediction probabilties by PRECOG.
                </p>
                <p class="text-justify fs-0"> Column information: </p>
                <ul>
                    <li class="text-justify fs-0"> Uniq_ID: A unique identifier assigned to your search by Precog</li>
                    <li class="text-justify fs-0"> GPCR: Given entry name</li>
                    <li class="text-justify fs-0"> VAR: Amino acid positional information of the variant (e.g. D154R)</li>
                    <li class="text-justify fs-0">P(WT): predicted probabilities of the Wild type sequence of the input</li>
                    <li class="text-justify fs-0">Mutations {eg: P(A263I)}: Predicted probabilites of the queried mutation (if provided)</li>
                    <li class="text-justify fs-0">IUPHAR: Information known from IUPHAR</li>
                    <li class="text-justify fs-0">LogRAi: Values known from the shedding assay. (Non-coupling GPCRs < -1.0 < Coupling GPCRs)</li>
                    <li class="text-justify fs-0">[GNAI3 - GNAL]: Probabilities (in case of Wild type and mutations) or known information
                        (in case of IUPHAR and LogRAi) of the entry for the corresponding G-protein
                        7TM1_POS/BW/ALN_POS: Values between '/' (Only applicable in case of mutations)</li>
                    <li class="text-justify fs-0">1st: The 7TM1-domain position of the mutation</li>
                    <li class="text-justify fs-0"> 2nd: Ballesteros-Weinstein numbering of the position affected</li>
                    <li class="text-justify fs-0"> 3rd: Internal numbering used by Precog</li>
                    <li class="text-justify fs-0"> [4th:]: The G-proteins(if any) affected by the mutation</li>
                    <li class="text-justify fs-0">Mutation_Info: Disease related information(if any) of the mutation retrieved from Uniprot</li>
                </ul>
                <h2 class="fw-bold mb-4">Browser compatibility</h2>
                <ul>
                    <li class="text-justify fs-0"> OS    Version    Chrome    Firefox    Safari</li>
                    <li class="text-justify fs-0"> Linux    Ubuntu 16.04    not tested    Quantum 64.0    n/a</li>
                    <li class="text-justify fs-0"> Linux    CentOS 7.2.1511    not tested    ESR 52.6.0    n/a</li>
                    <li class="text-justify fs-0"> Windows    10    v71    not tested    n/a</li>
                    <li class="text-justify fs-0"> MacOS    Mojave 10.14.2    v71    v64.0    12.0</li>
                </ul>
                <h5 class="fw-bold mb-4">Contacts</h5>
                <p class="text-justify fs-0">Francesco Raimondi - francesco.raimondi@sns.it</p>
                <p class="text-justify fs-0">Marin Matic - marin.matic@sns.it</p>
                <p class="text-justify fs-0">Gurdeep Sing - gurdeep.sing@bioquant.uni-heidelberg.de</p>
            </div>
        </div>
    </div>
</section>
{% endblock%}
